File: src/codegen.c
/* The Stańczyk Programming Language
 *
 *            ¿«fº"└└-.`└└*∞▄_              ╓▄∞╙╙└└└╙╙*▄▄
 *         J^. ,▄▄▄▄▄▄_      └▀████▄ç    JA▀            └▀v
 *       ,┘ ▄████████████▄¿     ▀██████▄▀└      ╓▄██████▄¿ "▄_
 *      ,─╓██▀└└└╙▀█████████      ▀████╘      ▄████████████_`██▄
 *     ;"▄█└      ,██████████-     ▐█▀      ▄███████▀▀J█████▄▐▀██▄
 *     ▌█▀      _▄█▀▀█████████      █      ▄██████▌▄▀╙     ▀█▐▄,▀██▄
 *    ▐▄▀     A└-▀▌  █████████      ║     J███████▀         ▐▌▌╙█µ▀█▄
 *  A╙└▀█∩   [    █  █████████      ▌     ███████H          J██ç ▀▄╙█_
 * █    ▐▌    ▀▄▄▀  J█████████      H    ████████          █    █  ▀▄▌
 *  ▀▄▄█▀.          █████████▌           ████████          █ç__▄▀ ╓▀└ ╙%_
 *                 ▐█████████      ▐    J████████▌          .└╙   █¿   ,▌
 *                 █████████▀╙╙█▌└▐█╙└██▀▀████████                 ╙▀▀▀▀
 *                ▐██▀┘Å▀▄A └▓█╓▐█▄▄██▄J▀@└▐▄Å▌▀██▌
 *                █▄▌▄█M╨╙└└-           .└└▀**▀█▄,▌
 *                ²▀█▄▄L_                  _J▄▄▄█▀└
 *                     └╙▀▀▀▀▀MMMR████▀▀▀▀▀▀▀└
 *
 *
 * ███████╗████████╗ █████╗ ███╗   ██╗ ██████╗███████╗██╗   ██╗██╗  ██╗
 * ██╔════╝╚══██╔══╝██╔══██╗████╗  ██║██╔════╝╚══███╔╝╚██╗ ██╔╝██║ ██╔╝
 * ███████╗   ██║   ███████║██╔██╗ ██║██║       ███╔╝  ╚████╔╝ █████╔╝
 * ╚════██║   ██║   ██╔══██║██║╚██╗██║██║      ███╔╝    ╚██╔╝  ██╔═██╗
 * ███████║   ██║   ██║  ██║██║ ╚████║╚██████╗███████╗   ██║   ██║  ██╗
 * ╚══════╝   ╚═╝   ╚═╝  ╚═╝╚═╝  ╚═══╝ ╚═════╝╚══════╝   ╚═╝   ╚═╝  ╚═╝
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "stanczyk.h"

#include "codegen.h"

#include "assembly.h"
#include "ir_code.h"
#include "logger.h"
#include "memory.h"

typedef struct {
    IRCodeChunk *chunk;
    Code *ip;
    Assembly *assembly;
} Codegen;

Codegen *codegen;

static void start_codegen(IRCodeChunk *chunk, Assembly *assembly) {
    codegen = ALLOCATE(Codegen);
    codegen->chunk = chunk;
    codegen->ip = chunk->code;
    codegen->assembly = assembly;
}

static void stop_codegen() {
    FREE(Codegen, codegen);
}

static char *str_ascii(String *str) {
    if (str->length > 255) {
        CODEGEN_ERROR("String has more than 255 characters. Split the string");
    }
    char *result = ALLOCATE_AMOUNT(char, 255);
    char *num = ALLOCATE_AMOUNT(char, 1);

    for (char c = *str->chars; c; c = *++str->chars) {
        if (c == '\\') {
            c = *++str->chars;
            switch (c) {
                case 't': sprintf(num, "%d", 9);  break;
                case 'n': sprintf(num, "%d", 10); break;
            }
        } else {
            sprintf(num, "%u", (u8)c);
        }

        strcat(result, num);
        strcat(result, ",");
    }
    strcat(result, "0");
    FREE(char, num);
    return result;
}

static void generate_linux_x86() {
#define NEXT_BYTE() (*codegen->ip++);
    AssemblyChunk *text = &codegen->assembly->text;
    AssemblyChunk *data = &codegen->assembly->data;

    write_assembly(text, "section .text");
    write_assembly(text, "global _start");
    write_assembly(text, "print:");
    write_assembly(text, "    mov     r9, -3689348814741910323");
    write_assembly(text, "    sub     rsp, 40");
    write_assembly(text, "    mov     BYTE [rsp+31], 10");
    write_assembly(text, "    lea     rcx, [rsp+30]");
    write_assembly(text, ".L2:");
    write_assembly(text, "    mov     rax, rdi");
    write_assembly(text, "    lea     r8, [rsp+32]");
    write_assembly(text, "    mul     r9");
    write_assembly(text, "    mov     rax, rdi");
    write_assembly(text, "    sub     r8, rcx");
    write_assembly(text, "    shr     rdx, 3");
    write_assembly(text, "    lea     rsi, [rdx+rdx*4]");
    write_assembly(text, "    add     rsi, rsi");
    write_assembly(text, "    sub     rax, rsi");
    write_assembly(text, "    add     eax, 48");
    write_assembly(text, "    mov     BYTE [rcx], al");
    write_assembly(text, "    mov     rax, rdi");
    write_assembly(text, "    mov     rdi, rdx");
    write_assembly(text, "    mov     rdx, rcx");
    write_assembly(text, "    sub     rcx, 1");
    write_assembly(text, "    cmp     rax, 9");
    write_assembly(text, "    ja      .L2");
    write_assembly(text, "    lea     rax, [rsp+32]");
    write_assembly(text, "    mov     edi, 1");
    write_assembly(text, "    sub     rdx, rax");
    write_assembly(text, "    xor     eax, eax");
    write_assembly(text, "    lea     rsi, [rsp+32+rdx]");
    write_assembly(text, "    mov     rdx, r8");
    write_assembly(text, "    mov     rax, 1");
    write_assembly(text, "    syscall");
    write_assembly(text, "    add     rsp, 40");
    write_assembly(text, "    ret");
    write_assembly(text, "_start:");
    write_assembly(text, ";; user program definitions starts here:");

    write_assembly(data, "section .data");

    for (;;) {
        int offset = codegen->ip - codegen->chunk->code;
        Code instruction = NEXT_BYTE();
        Token *token = &instruction.token;

        write_assembly(text, "ip_%d:", offset);

        switch (instruction.type) {
            /*    ___             _            _
             *   / __|___ _ _  __| |_ __ _ _ _| |_ ___
             *  | (__/ _ \ ' \(_-<  _/ _` | ' \  _(_-<
             *   \___\___/_||_/__/\__\__,_|_||_\__/__/
             */
            case OP_PUSH_INT: {
                int value = AS_INT(instruction.operand);
                write_assembly(text, ";; %d (%s:%d:%d)", value,
                               token->filename, token->line, token->column);
                write_assembly(text, "    mov rax, %d", value);
                write_assembly(text, "    push rax");
            } break;
            case OP_PUSH_STR: {
                String *value = AS_STRING(instruction.operand);
                char *ascii_str = str_ascii(value);
                write_assembly(text, ";; %s (%s:%d:%d)", value->chars,
                               token->filename, token->line, token->column);
                write_assembly(text, "    mov rax, %d", value->length);
                write_assembly(text, "    push rax");
                write_assembly(text, "    push str_%d", data->count);
                write_assembly(data, "str_%d: db %s", data->count, ascii_str);
            } break;
            case OP_PUSH_BOOL: {
                int value = AS_INT(instruction.operand);
                write_assembly(text, ";; %.*s (%s:%d:%d)", token->length, token->start,
                               token->filename, token->line, token->column);
                write_assembly(text, "    mov rax, %d", value);
                write_assembly(text, "    push rax");
            } break;

            /*   ___     _       _         _
             *  |_ _|_ _| |_ _ _(_)_ _  __(_)__ ___
             *   | || ' \  _| '_| | ' \(_-< / _(_-<
             *  |___|_||_\__|_| |_|_||_/__/_\__/__/
             */
            case OP_ADD: {
                write_assembly(text, ";; + (%s:%d:%d)",
                               token->filename, token->line, token->column);
                write_assembly(text, "    pop rax");
                write_assembly(text, "    pop rbx");
                write_assembly(text, "    add rax, rbx");
                write_assembly(text, "    push rax");
            } break;
            case OP_SUBSTRACT: {
                write_assembly(text, ";; - (%s:%d:%d)",
                               token->filename, token->line, token->column);
                write_assembly(text, "    pop rbx");
                write_assembly(text, "    pop rax");
                write_assembly(text, "    sub rax, rbx");
                write_assembly(text, "    push rax");
            } break;
            case OP_MULTIPLY: {
                write_assembly(text, ";; * (%s:%d:%d)",
                               token->filename, token->line, token->column);
                write_assembly(text, "    pop rax");
                write_assembly(text, "    pop rbx");
                write_assembly(text, "    mul rbx");
                write_assembly(text, "    push rax");
            } break;
            case OP_DIVIDE: {
                write_assembly(text, ";; / (%s:%d:%d)",
                               token->filename, token->line, token->column);
                write_assembly(text, "    xor rdx, rdx");
                write_assembly(text, "    pop rbx");
                write_assembly(text, "    pop rax");
                write_assembly(text, "    div rbx");
                write_assembly(text, "    push rax");
            } break;
            case OP_MODULO: {
                write_assembly(text, ";; % (%s:%d:%d)",
                               token->filename, token->line, token->column);
                write_assembly(text, "    xor rdx, rdx");
                write_assembly(text, "    pop rbx");
                write_assembly(text, "    pop rax");
                write_assembly(text, "    div rbx");
                write_assembly(text, "    push rdx");
            } break;
            case OP_EQUAL: {
                write_assembly(text, ";; = (%s:%d:%d)",
                               token->filename, token->line, token->column);
                write_assembly(text, "    xor rcx, rcx");
                write_assembly(text, "    mov rdx, 1");
                write_assembly(text, "    pop rax");
                write_assembly(text, "    pop rbx");
                write_assembly(text, "    cmp rax, rbx");
                write_assembly(text, "    cmove rcx, rdx");
                write_assembly(text, "    push rcx");
            } break;
            case OP_NOT_EQUAL: {
                write_assembly(text, ";; != (%s:%d:%d)",
                               token->filename, token->line, token->column);
                write_assembly(text, "    xor rcx, rcx");
                write_assembly(text, "    mov rdx, 1");
                write_assembly(text, "    pop rax");
                write_assembly(text, "    pop rbx");
                write_assembly(text, "    cmp rax, rbx");
                write_assembly(text, "    cmovne rcx, rdx");
                write_assembly(text, "    push rcx");
            } break;
            case OP_LESS: {
                write_assembly(text, ";; < (%s:%d:%d)",
                               token->filename, token->line, token->column);
                write_assembly(text, "    xor rcx, rcx");
                write_assembly(text, "    mov rdx, 1");
                write_assembly(text, "    pop rbx");
                write_assembly(text, "    pop rax");
                write_assembly(text, "    cmp rax, rbx");
                write_assembly(text, "    cmovl rcx, rdx");
                write_assembly(text, "    push rcx");
            } break;
            case OP_LESS_EQUAL: {
                write_assembly(text, ";; <= (%s:%d:%d)",
                               token->filename, token->line, token->column);
                write_assembly(text, "    xor rcx, rcx");
                write_assembly(text, "    mov rdx, 1");
                write_assembly(text, "    pop rbx");
                write_assembly(text, "    pop rax");
                write_assembly(text, "    cmp rax, rbx");
                write_assembly(text, "    cmovle rcx, rdx");
                write_assembly(text, "    push rcx");
            } break;
            case OP_GREATER: {
                write_assembly(text, ";; > (%s:%d:%d)",
                               token->filename, token->line, token->column);
                write_assembly(text, "    xor rcx, rcx");
                write_assembly(text, "    mov rdx, 1");
                write_assembly(text, "    pop rbx");
                write_assembly(text, "    pop rax");
                write_assembly(text, "    cmp rax, rbx");
                write_assembly(text, "    cmovg rcx, rdx");
                write_assembly(text, "    push rcx");
            } break;
            case OP_GREATER_EQUAL: {
                write_assembly(text, ";; >= (%s:%d:%d)",
                               token->filename, token->line, token->column);
                write_assembly(text, "    xor rcx, rcx");
                write_assembly(text, "    mov rdx, 1");
                write_assembly(text, "    pop rbx");
                write_assembly(text, "    pop rax");
                write_assembly(text, "    cmp rax, rbx");
                write_assembly(text, "    cmovge rcx, rdx");
                write_assembly(text, "    push rcx");
            } break;
            case OP_DROP: {
                write_assembly(text, ";; drop (%s:%d:%d)",
                               token->filename, token->line, token->column);
                write_assembly(text, "    pop rax");
            } break;
            case OP_DUP: {
                write_assembly(text, ";; dup (%s:%d:%d)",
                               token->filename, token->line, token->column);
                write_assembly(text, "    pop rax");
                write_assembly(text, "    push rax");
                write_assembly(text, "    push rax");
            } break;
            case OP_PRINT: {
                write_assembly(text, ";; print (%s:%d:%d)",
                               token->filename, token->line, token->column);
                write_assembly(text, "    pop rdi");
                write_assembly(text, "    call print");
            } break;
            case OP_JUMP_IF_FALSE: {
                int value = AS_INT(instruction.operand);
                write_assembly(text, ";; do (%s:%d:%d)",
                               token->filename, token->line, token->column);
                write_assembly(text, "    pop rax");
                write_assembly(text, "    test rax, rax");
                write_assembly(text, "    jz ip_%d", value);
            } break;
            case OP_JUMP: {
                int value = AS_INT(instruction.operand);
                write_assembly(text, ";; else (%s:%d:%d)",
                               token->filename, token->line, token->column);
                write_assembly(text, "    jmp ip_%d", value);
            } break;
            case OP_LOOP: {
                int value = AS_INT(instruction.operand);
                write_assembly(text, ";; loop (%s:%d:%d)",
                               token->filename, token->line, token->column);
                write_assembly(text, "    jmp ip_%d", value);
            } break;
            case OP_SYSCALL3: {
                write_assembly(text, ";; __SYSCALL3 (%s:%d:%d)",
                               token->filename, token->line, token->column);
                write_assembly(text, "    pop rax");
                write_assembly(text, "    pop rdi");
                write_assembly(text, "    pop rsi");
                write_assembly(text, "    pop rdx");
                write_assembly(text, "    syscall");
                write_assembly(text, "    push rax");
            } break;


            /*   _  __                           _
             *  | |/ /___ _  ___ __ _____ _ _ __| |___
             *  | ' </ -_) || \ V  V / _ \ '_/ _` (_-<
             *  |_|\_\___|\_, |\_/\_/\___/_| \__,_/__/
             *            |__/
             */
            case OP_CAST:
            case OP_END_IF:
            case OP_END_LOOP: break;

            case OP_EOC: {
                write_assembly(text, ";; eof (%s:%d:%d)",
                               token->filename, token->line, token->column);
                write_assembly(text, "    mov rax, 60");
                write_assembly(text, "    mov rdi, 0");
                write_assembly(text, "    syscall");
            } return;
        }
    }
#undef NEXT_BYTE
}

void codegen_run(IRCodeChunk *chunk, Assembly *assembly) {
    start_codegen(chunk, assembly);

#if defined(__linux__)
    generate_linux_x86();
#elif defined(_WIN32) || defined(_WIN64) || defined(__CYGWIN__)
    CODEGEN_ERROR("Sorry, this Operating System is not supported at this moment");
#elif defined(__APPLE__) || defined(__MACH__)
    CODEGEN_ERROR("Sorry, this Operating System is not supported at this moment");
#else
    CODEGEN_ERROR("Sorry, this Operating System is not supported at this moment");
#endif

    stop_codegen();
}

File: code/codegen.go
package main

import (
	"fmt"
	"runtime"
)

type Codegen struct {
	asm *Assembly
}

var codegen Codegen

func getAsciiValues(s string) (string, int) {
	count := 0
	nextEscaped := false
	ascii := ""

	for _, c := range s {
		if c == '\\' {
			nextEscaped = true
			continue
		}

		if nextEscaped {
			switch c {
			case 't':
				ascii += "9,"
				count++
			case 'n':
				ascii += "10,"
				count++
			case 'e':
				ascii += "27,"
				count++
			}

			nextEscaped = false
			continue
		}

		val := fmt.Sprintf("%v,", c)
		ascii += val
		count++
	}

	ascii += "0"
	count++

	return ascii, count
}

func generateLinuxX86() {
	mainFuncIP := -1
	asm := codegen.asm

	asm.WriteText("section .text")
    asm.WriteText("global _start")
    asm.WriteText("print:")
    asm.WriteText("    mov     r9, -3689348814741910323")
    asm.WriteText("    sub     rsp, 40")
    asm.WriteText("    mov     BYTE [rsp+31], 10")
    asm.WriteText("    lea     rcx, [rsp+30]")
    asm.WriteText(".L2:")
    asm.WriteText("    mov     rax, rdi")
    asm.WriteText("    lea     r8, [rsp+32]")
    asm.WriteText("    mul     r9")
    asm.WriteText("    mov     rax, rdi")
    asm.WriteText("    sub     r8, rcx")
    asm.WriteText("    shr     rdx, 3")
    asm.WriteText("    lea     rsi, [rdx+rdx*4]")
    asm.WriteText("    add     rsi, rsi")
    asm.WriteText("    sub     rax, rsi")
    asm.WriteText("    add     eax, 48")
    asm.WriteText("    mov     BYTE [rcx], al")
    asm.WriteText("    mov     rax, rdi")
    asm.WriteText("    mov     rdi, rdx")
    asm.WriteText("    mov     rdx, rcx")
    asm.WriteText("    sub     rcx, 1")
    asm.WriteText("    cmp     rax, 9")
    asm.WriteText("    ja      .L2")
    asm.WriteText("    lea     rax, [rsp+32]")
    asm.WriteText("    mov     edi, 1")
    asm.WriteText("    sub     rdx, rax")
    asm.WriteText("    xor     eax, eax")
    asm.WriteText("    lea     rsi, [rsp+32+rdx]")
    asm.WriteText("    mov     rdx, r8")
    asm.WriteText("    mov     rax, 1")
    asm.WriteText("    syscall")
    asm.WriteText("    add     rsp, 40")
    asm.WriteText("    ret")
    asm.WriteText(";; user program definitions starts here")

	asm.WriteData("section .data")

	asm.WriteBss("section .bss")
	asm.WriteBss("args_ptr: resq 1")
	asm.WriteBss("return_stack_rsp: resq 1")
	asm.WriteBss("return_stack: resb 65536")
	asm.WriteBss("return_stack_rsp_end:")

	for _, function := range TheProgram.chunks {
		if !function.called {
			continue
		}
		if function.name == "main" {
			mainFuncIP = function.ip
		}

		asm.WriteText("fn%d:", function.ip)
		asm.WriteText(";; start function %s (%s:%d:%d)", function.name,
			function.loc.f, function.loc.l, function.loc.c)
		asm.WriteText("    sub rsp, 0")
		asm.WriteText("    mov [return_stack_rsp], rsp")
		asm.WriteText("    mov rsp, rax")

		for index, code := range function.code {
			instruction := code.op
			loc := code.loc
			value := code.value

			asm.WriteText(".ip%d:", index)

			switch instruction {
			// Constants
			case OP_PUSH_BOOL:
				boolText := map[int]string{1:"true", 0:"false"} [value.(int)]
				asm.WriteText(";; %s (%s:%d:%d)", boolText, loc.f, loc.l, loc.c)
				asm.WriteText("    mov rax, %d", value)
				asm.WriteText("    push rax")
			case OP_PUSH_INT:
				asm.WriteText(";; %d (%s:%d:%d)", value, loc.f, loc.l, loc.c)
				asm.WriteText("    mov rax, %d", value)
				asm.WriteText("    push rax")
			case OP_PUSH_STR:
				ascii, _ := getAsciiValues(value.(string))
				asm.WriteText(";; \"%s\" (%s:%d:%d)", value, loc.f, loc.l, loc.c)
				asm.WriteText("    push str_%d", len(asm.data))
				asm.WriteData("str_%d: db %s", len(asm.data), ascii)

				// Intrinsics
			case OP_ADD:
				asm.WriteText(";; + (%s:%d:%d)", loc.f, loc.l, loc.c)
				asm.WriteText("    pop rax")
				asm.WriteText("    pop rbx")
				asm.WriteText("    add rax, rbx")
				asm.WriteText("    push rax")
			case OP_ARGC:
				asm.WriteText(";; argc (%s:%d:%d)", loc.f, loc.l, loc.c)
				asm.WriteText("    mov rax, [args_ptr]")
                asm.WriteText("    mov rax, [rax]")
                asm.WriteText("    push rax")
			case OP_ARGV:
				asm.WriteText(";; argv (%s:%d:%d)", loc.f, loc.l, loc.c)
				asm.WriteText("    mov rax, [args_ptr]")
				asm.WriteText("    add rax, 8")
                asm.WriteText("    push rax")
			case OP_CAST:
				asm.WriteText(";; cast to %s (%s:%d:%d)",
					getDataTypeName(value.(DataType)), loc.f, loc.l, loc.c)
			case OP_DIVIDE:
				asm.WriteText(";; / (%s:%d:%d)", loc.f, loc.l, loc.c)
				asm.WriteText("    xor rdx, rdx")
				asm.WriteText("    pop rbx")
				asm.WriteText("    pop rax")
				asm.WriteText("    div rbx")
				asm.WriteText("    push rax")
				asm.WriteText("    push rdx")
			case OP_DROP:
				asm.WriteText(";; drop (%s:%d:%d)", loc.f, loc.l, loc.c)
				asm.WriteText("    pop rax")
			case OP_DUP:
				asm.WriteText(";; dup (%s:%d:%d)", loc.f, loc.l, loc.c)
				asm.WriteText("    pop rax")
				asm.WriteText("    push rax")
				asm.WriteText("    push rax")
			case OP_EQUAL:
				asm.WriteText(";; = (%s:%d:%d)", loc.f, loc.l, loc.c)
				asm.WriteText("    xor rcx, rcx")
				asm.WriteText("    mov rdx, 1")
				asm.WriteText("    pop rax")
				asm.WriteText("    pop rbx")
				asm.WriteText("    cmp rax, rbx")
				asm.WriteText("    cmove rcx, rdx")
				asm.WriteText("    push rcx")
			case OP_GREATER:
				asm.WriteText(";; > (%s:%d:%d)", loc.f, loc.l, loc.c)
				asm.WriteText("    xor rcx, rcx")
				asm.WriteText("    mov rdx, 1")
				asm.WriteText("    pop rbx")
				asm.WriteText("    pop rax")
				asm.WriteText("    cmp rax, rbx")
				asm.WriteText("    cmovg rcx, rdx")
				asm.WriteText("    push rcx")
			case OP_GREATER_EQUAL:
				asm.WriteText(";; >= (%s:%d:%d)", loc.f, loc.l, loc.c)
				asm.WriteText("    xor rcx, rcx")
				asm.WriteText("    mov rdx, 1")
				asm.WriteText("    pop rbx")
				asm.WriteText("    pop rax")
				asm.WriteText("    cmp rax, rbx")
				asm.WriteText("    cmovge rcx, rdx")
				asm.WriteText("    push rcx")
			case OP_LESS:
				asm.WriteText(";; < (%s:%d:%d)", loc.f, loc.l, loc.c)
				asm.WriteText("    xor rcx, rcx")
				asm.WriteText("    mov rdx, 1")
				asm.WriteText("    pop rbx")
				asm.WriteText("    pop rax")
				asm.WriteText("    cmp rax, rbx")
				asm.WriteText("    cmovl rcx, rdx")
				asm.WriteText("    push rcx")
			case OP_LESS_EQUAL:
				asm.WriteText(";; <= (%s:%d:%d)", loc.f, loc.l, loc.c)
				asm.WriteText("    xor rcx, rcx")
				asm.WriteText("    mov rdx, 1")
				asm.WriteText("    pop rbx")
				asm.WriteText("    pop rax")
				asm.WriteText("    cmp rax, rbx")
				asm.WriteText("    cmovle rcx, rdx")
				asm.WriteText("    push rcx")
			case OP_LOAD8:
				asm.WriteText("    pop rax")
                asm.WriteText("    xor rbx, rbx")
                asm.WriteText("    mov bl, [rax]")
                asm.WriteText("    push rbx")
			case OP_LOAD16:
				asm.WriteText("    pop rax")
                asm.WriteText("    xor rbx, rbx")
                asm.WriteText("    mov bx, [rax]")
                asm.WriteText("    push rbx")
			case OP_LOAD32:
				asm.WriteText("    pop rax")
                asm.WriteText("    xor rbx, rbx")
                asm.WriteText("    mov ebx, [rax]")
                asm.WriteText("    push rbx")
			case OP_LOAD64:
				asm.WriteText("    pop rax")
                asm.WriteText("    xor rbx, rbx")
                asm.WriteText("    mov rbx, [rax]")
                asm.WriteText("    push rbx")
			case OP_MULTIPLY:
				asm.WriteText(";; * (%s:%d:%d)", loc.f, loc.l, loc.c)
				asm.WriteText("    pop rax")
				asm.WriteText("    pop rbx")
				asm.WriteText("    mul rbx")
				asm.WriteText("    push rax")
			case OP_NOT_EQUAL:
				asm.WriteText(";; != (%s:%d:%d)", loc.f, loc.l, loc.c)
				asm.WriteText("    xor rcx, rcx")
				asm.WriteText("    mov rdx, 1")
				asm.WriteText("    pop rax")
				asm.WriteText("    pop rbx")
				asm.WriteText("    cmp rax, rbx")
				asm.WriteText("    cmovne rcx, rdx")
				asm.WriteText("    push rcx")
			case OP_OVER:
				asm.WriteText(";; over (%s:%d:%d)", loc.f, loc.l, loc.c)
				asm.WriteText("    pop rax")
				asm.WriteText("    pop rbx")
				asm.WriteText("    push rbx")
				asm.WriteText("    push rax")
				asm.WriteText("    push rbx")
			case OP_PRINT:
				asm.WriteText(";; print (%s:%d:%d)", loc.f, loc.l, loc.c)
				asm.WriteText("    pop rdi")
				asm.WriteText("    call print")
			case OP_SUBSTRACT:
				asm.WriteText(";; - (%s:%d:%d)", loc.f, loc.l, loc.c)
				asm.WriteText("    pop rbx")
				asm.WriteText("    pop rax")
				asm.WriteText("    sub rax, rbx")
				asm.WriteText("    push rax")
			case OP_STORE8:
                asm.WriteText("    pop rax")
                asm.WriteText("    pop rbx")
                asm.WriteText("    mov [rax], bl")
			case OP_STORE16:
                asm.WriteText("    pop rax")
                asm.WriteText("    pop rbx")
                asm.WriteText("    mov [rax], bx")
			case OP_STORE32:
                asm.WriteText("    pop rax")
                asm.WriteText("    pop rbx")
                asm.WriteText("    mov [rax], ebx")
			case OP_STORE64:
                asm.WriteText("    pop rax")
                asm.WriteText("    pop rbx")
                asm.WriteText("    mov [rax], rbx")
			case OP_SWAP:
				asm.WriteText(";; swap (%s:%d:%d)", loc.f, loc.l, loc.c)
				asm.WriteText("    pop rax")
				asm.WriteText("    pop rbx")
				asm.WriteText("    push rax")
				asm.WriteText("    push rbx")
			case OP_RET:
				asm.WriteText(";; ret (%s:%d:%d)", loc.f, loc.l, loc.c)
				asm.WriteText("    mov rax, rsp")
				asm.WriteText("    mov rsp, [return_stack_rsp]")
				asm.WriteText("    add rsp, %d", value)
				asm.WriteText("    ret")

				// Special
			case OP_SYSCALL:
				regs := []string{"rax", "rdi", "rsi", "rdx", "r10", "r8", "r9",}
				for i, _ := range function.args {
					asm.WriteText("    pop %s", regs[i])
				}
				asm.WriteText("    syscall")
				for range function.rets {
					asm.WriteText("    push rax")
				}
			case OP_WORD:
				fnCall := FindFunction(code)

				asm.WriteText(";; call function %s (%s:%d:%d)", fnCall.name, loc.f, loc.l, loc.c)
				asm.WriteText("    mov rax, rsp")
				asm.WriteText("    mov rsp, [return_stack_rsp]")
				asm.WriteText("    call fn%d", fnCall.ip)
				asm.WriteText("    mov [return_stack_rsp], rsp")
				asm.WriteText("    mov rsp, rax")
			case OP_END_IF:
				asm.WriteText(";; . [if] (%s:%d:%d)", loc.f, loc.l, loc.c)
			case OP_END_LOOP:
				asm.WriteText(";; . [loop] (%s:%d:%d)", loc.f, loc.l, loc.c)
			case OP_IF:
				asm.WriteText(";; if (%s:%d:%d)", loc.f, loc.l, loc.c)
			case OP_JUMP:
				asm.WriteText(";; else (%s:%d:%d)", loc.f, loc.l, loc.c)
				asm.WriteText("    jmp .ip%d", value)
			case OP_JUMP_IF_FALSE:
				asm.WriteText(";; do (%s:%d:%d)", loc.f, loc.l, loc.c)
				asm.WriteText("    pop rax")
				asm.WriteText("    test rax, rax")
				asm.WriteText("    jz .ip%d", value)
			case OP_LOOP:
				asm.WriteText(";; loop (%s:%d:%d)", loc.f, loc.l, loc.c)
				asm.WriteText("    jmp .ip%d", value)
			}
		}

		asm.WriteText(";; end function %s (%s:%d:%d)", function.name,
			function.loc.f, function.loc.l, function.loc.c)
	}

	asm.WriteText(";; user program definition ends here")
	asm.WriteText("_start:")
	asm.WriteText("    mov [args_ptr], rsp")
	asm.WriteText("    mov rax, return_stack_rsp_end")
	asm.WriteText("    mov [return_stack_rsp], rax")
	asm.WriteText("    call fn%d", mainFuncIP)
	asm.WriteText("    mov rax, 60")
	asm.WriteText("    mov rdi, 0")
	asm.WriteText("    syscall")
}

func CodegenRun(asm *Assembly) {
	codegen.asm = asm

	switch runtime.GOOS {
	case "linux":
		generateLinuxX86()
	default:
		Stanczyk.Error("OS currently not supported")
	}
}
